File: sdks/go/heliosdb.go
// Package heliosdb provides a Go client for HeliosDB.
//
// Example usage:
//
//	client, err := heliosdb.NewClient("http://localhost:8080",
//	    heliosdb.WithAPIKey("your-api-key"),
//	    heliosdb.WithBranch("main"),
//	)
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	// Execute SQL query
//	result, err := client.Query(ctx, "SELECT * FROM users WHERE id = $1", 1)
//
//	// Vector search
//	results, err := client.VectorSearch(ctx, "documents", "hello world",
//	    heliosdb.WithTopK(10),
//	    heliosdb.WithMinScore(0.7),
//	)
package heliosdb

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

// Client is the HeliosDB client.
type Client struct {
	baseURL    string
	apiKey     string
	branch     string
	httpClient *http.Client
}

// ClientOption configures the client.
type ClientOption func(*Client)

// WithAPIKey sets the API key for authentication.
func WithAPIKey(key string) ClientOption {
	return func(c *Client) {
		c.apiKey = key
	}
}

// WithBranch sets the default branch.
func WithBranch(branch string) ClientOption {
	return func(c *Client) {
		c.branch = branch
	}
}

// WithHTTPClient sets a custom HTTP client.
func WithHTTPClient(client *http.Client) ClientOption {
	return func(c *Client) {
		c.httpClient = client
	}
}

// WithTimeout sets the default timeout.
func WithTimeout(d time.Duration) ClientOption {
	return func(c *Client) {
		c.httpClient.Timeout = d
	}
}

// NewClient creates a new HeliosDB client.
func NewClient(baseURL string, opts ...ClientOption) (*Client, error) {
	if _, err := url.Parse(baseURL); err != nil {
		return nil, fmt.Errorf("invalid base URL: %w", err)
	}

	c := &Client{
		baseURL: baseURL,
		branch:  "main",
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}

	for _, opt := range opts {
		opt(c)
	}

	return c, nil
}

// QueryResult represents the result of a SQL query.
type QueryResult struct {
	Rows        []map[string]interface{} `json:"rows"`
	Columns     []string                 `json:"columns"`
	RowsAffected int64                   `json:"rows_affected,omitempty"`
}

// Query executes a SQL query.
func (c *Client) Query(ctx context.Context, sql string, params ...interface{}) (*QueryResult, error) {
	return c.QueryOnBranch(ctx, c.branch, sql, params...)
}

// QueryOnBranch executes a SQL query on a specific branch.
func (c *Client) QueryOnBranch(ctx context.Context, branch, sql string, params ...interface{}) (*QueryResult, error) {
	body := map[string]interface{}{
		"sql":    sql,
		"params": params,
	}

	var result QueryResult
	err := c.post(ctx, fmt.Sprintf("/v1/branches/%s/query", branch), body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Exec executes a SQL statement that doesn't return rows.
func (c *Client) Exec(ctx context.Context, sql string, params ...interface{}) (int64, error) {
	result, err := c.Query(ctx, sql, params...)
	if err != nil {
		return 0, err
	}
	return result.RowsAffected, nil
}

// VectorStore represents a vector store.
type VectorStore struct {
	Name       string `json:"name"`
	Dimensions int    `json:"dimensions"`
	Metric     string `json:"metric"`
	Count      int64  `json:"count"`
}

// VectorSearchResult represents a single search result.
type VectorSearchResult struct {
	ID       string                 `json:"id"`
	Score    float64                `json:"score"`
	Content  string                 `json:"content,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// VectorSearchOptions configures vector search.
type VectorSearchOptions struct {
	TopK     int
	MinScore float64
	Filter   map[string]interface{}
}

// VectorSearchOption configures vector search.
type VectorSearchOption func(*VectorSearchOptions)

// WithTopK sets the maximum number of results.
func WithTopK(k int) VectorSearchOption {
	return func(o *VectorSearchOptions) {
		o.TopK = k
	}
}

// WithMinScore sets the minimum similarity score.
func WithMinScore(score float64) VectorSearchOption {
	return func(o *VectorSearchOptions) {
		o.MinScore = score
	}
}

// WithFilter sets metadata filter.
func WithFilter(filter map[string]interface{}) VectorSearchOption {
	return func(o *VectorSearchOptions) {
		o.Filter = filter
	}
}

// VectorSearch performs semantic search by text.
func (c *Client) VectorSearch(ctx context.Context, store, query string, opts ...VectorSearchOption) ([]VectorSearchResult, error) {
	options := &VectorSearchOptions{
		TopK: 10,
	}
	for _, opt := range opts {
		opt(options)
	}

	body := map[string]interface{}{
		"text":   query,
		"top_k":  options.TopK,
	}
	if options.MinScore > 0 {
		body["min_score"] = options.MinScore
	}
	if options.Filter != nil {
		body["filter"] = options.Filter
	}

	var response struct {
		Results []VectorSearchResult `json:"results"`
	}

	err := c.post(ctx, fmt.Sprintf("/v1/vectors/stores/%s/search/text", store), body, &response)
	if err != nil {
		return nil, err
	}

	return response.Results, nil
}

// VectorSearchByVector performs search by vector.
func (c *Client) VectorSearchByVector(ctx context.Context, store string, vector []float32, opts ...VectorSearchOption) ([]VectorSearchResult, error) {
	options := &VectorSearchOptions{
		TopK: 10,
	}
	for _, opt := range opts {
		opt(options)
	}

	body := map[string]interface{}{
		"vector": vector,
		"top_k":  options.TopK,
	}
	if options.MinScore > 0 {
		body["min_score"] = options.MinScore
	}
	if options.Filter != nil {
		body["filter"] = options.Filter
	}

	var response struct {
		Results []VectorSearchResult `json:"results"`
	}

	err := c.post(ctx, fmt.Sprintf("/v1/vectors/stores/%s/search", store), body, &response)
	if err != nil {
		return nil, err
	}

	return response.Results, nil
}

// StoreText stores text with automatic embedding.
func (c *Client) StoreText(ctx context.Context, store, text string, metadata map[string]interface{}) (string, error) {
	body := map[string]interface{}{
		"texts":     []string{text},
		"metadatas": []map[string]interface{}{metadata},
	}

	var response struct {
		IDs []string `json:"ids"`
	}

	err := c.post(ctx, fmt.Sprintf("/v1/vectors/stores/%s/texts", store), body, &response)
	if err != nil {
		return "", err
	}

	if len(response.IDs) > 0 {
		return response.IDs[0], nil
	}
	return "", nil
}

// Branch represents a database branch.
type Branch struct {
	Name      string  `json:"name"`
	Parent    *string `json:"parent,omitempty"`
	CreatedAt string  `json:"created_at"`
}

// ListBranches returns all branches.
func (c *Client) ListBranches(ctx context.Context) ([]Branch, error) {
	var response struct {
		Branches []Branch `json:"branches"`
	}

	err := c.get(ctx, "/v1/branches", &response)
	if err != nil {
		return nil, err
	}

	return response.Branches, nil
}

// CreateBranch creates a new branch.
func (c *Client) CreateBranch(ctx context.Context, name, fromBranch string) (*Branch, error) {
	body := map[string]interface{}{
		"name":        name,
		"from_branch": fromBranch,
	}

	var branch Branch
	err := c.post(ctx, "/v1/branches", body, &branch)
	if err != nil {
		return nil, err
	}

	return &branch, nil
}

// MergeBranch merges a branch into another.
func (c *Client) MergeBranch(ctx context.Context, source, target string) error {
	body := map[string]interface{}{
		"target": target,
	}

	return c.post(ctx, fmt.Sprintf("/v1/branches/%s/merge", source), body, nil)
}

// SetBranch sets the default branch for queries.
func (c *Client) SetBranch(branch string) {
	c.branch = branch
}

// GetBranch returns the current default branch.
func (c *Client) GetBranch() string {
	return c.branch
}

// AgentMemory provides agent memory operations.
type AgentMemory struct {
	client    *Client
	sessionID string
}

// Memory returns an agent memory instance for the given session.
func (c *Client) Memory(sessionID string) *AgentMemory {
	return &AgentMemory{
		client:    c,
		sessionID: sessionID,
	}
}

// Message represents a chat message.
type Message struct {
	Role      string `json:"role"`
	Content   string `json:"content"`
	Timestamp string `json:"timestamp,omitempty"`
}

// Add adds a message to memory.
func (m *AgentMemory) Add(ctx context.Context, role, content string) error {
	body := map[string]interface{}{
		"role":    role,
		"content": content,
	}

	return m.client.post(ctx, fmt.Sprintf("/v1/agents/memory/%s/add", m.sessionID), body, nil)
}

// Get retrieves messages from memory.
func (m *AgentMemory) Get(ctx context.Context, limit int) ([]Message, error) {
	var response struct {
		Messages []Message `json:"messages"`
	}

	err := m.client.get(ctx, fmt.Sprintf("/v1/agents/memory/%s/messages?limit=%d", m.sessionID, limit), &response)
	if err != nil {
		return nil, err
	}

	return response.Messages, nil
}

// Search searches memory semantically.
func (m *AgentMemory) Search(ctx context.Context, query string, topK int) ([]VectorSearchResult, error) {
	body := map[string]interface{}{
		"query": query,
		"top_k": topK,
	}

	var response struct {
		Results []VectorSearchResult `json:"results"`
	}

	err := m.client.post(ctx, fmt.Sprintf("/v1/agents/memory/%s/search", m.sessionID), body, &response)
	if err != nil {
		return nil, err
	}

	return response.Results, nil
}

// Clear clears the memory.
func (m *AgentMemory) Clear(ctx context.Context) error {
	return m.client.delete(ctx, fmt.Sprintf("/v1/agents/memory/%s", m.sessionID))
}

// QueryAt queries data at a specific point in time.
func (c *Client) QueryAt(ctx context.Context, sql string, timestamp time.Time, params ...interface{}) (*QueryResult, error) {
	body := map[string]interface{}{
		"sql":             sql,
		"params":          params,
		"as_of_timestamp": timestamp.Format(time.RFC3339),
	}

	var result QueryResult
	err := c.post(ctx, fmt.Sprintf("/v1/branches/%s/query", c.branch), body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// NLQuery executes a natural language query.
func (c *Client) NLQuery(ctx context.Context, question string) (*QueryResult, string, error) {
	body := map[string]interface{}{
		"question": question,
		"branch":   c.branch,
	}

	var response struct {
		SQL    string                   `json:"sql"`
		Rows   []map[string]interface{} `json:"rows"`
		Columns []string                `json:"columns"`
	}

	err := c.post(ctx, "/v1/nl/query", body, &response)
	if err != nil {
		return nil, "", err
	}

	result := &QueryResult{
		Rows:    response.Rows,
		Columns: response.Columns,
	}

	return result, response.SQL, nil
}

// Health checks the server health.
func (c *Client) Health(ctx context.Context) error {
	return c.get(ctx, "/health", nil)
}

// HTTP helpers

func (c *Client) get(ctx context.Context, path string, result interface{}) error {
	req, err := http.NewRequestWithContext(ctx, "GET", c.baseURL+path, nil)
	if err != nil {
		return err
	}

	return c.do(req, result)
}

func (c *Client) post(ctx context.Context, path string, body, result interface{}) error {
	jsonBody, err := json.Marshal(body)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", c.baseURL+path, bytes.NewReader(jsonBody))
	if err != nil {
		return err
	}

	return c.do(req, result)
}

func (c *Client) delete(ctx context.Context, path string) error {
	req, err := http.NewRequestWithContext(ctx, "DELETE", c.baseURL+path, nil)
	if err != nil {
		return err
	}

	return c.do(req, nil)
}

func (c *Client) do(req *http.Request, result interface{}) error {
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("X-API-Key", c.apiKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return &APIError{
			StatusCode: resp.StatusCode,
			Message:    string(body),
		}
	}

	if result != nil {
		return json.NewDecoder(resp.Body).Decode(result)
	}

	return nil
}

// APIError represents an API error.
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("API error %d: %s", e.StatusCode, e.Message)
}

File: sdks/typescript/src/client/heliosdb.ts
/**
 * HeliosDB main client
 */

import { request } from 'undici';
import {
  QueryResult,
  Column,
  TableInfo,
  TableSchema,
  Branch,
  HealthStatus,
  InferredSchema,
  QueryOptions,
  InsertOptions,
  CreateBranchOptions,
  TimeTravelOptions,
} from '../types';
import { createErrorFromResponse, ValidationError } from '../errors';
import { VectorStore, VectorStoreConfig } from '../vector/store';
import { AgentMemory } from '../memory/agent-memory';
import { BranchContext, TableQuery } from './branch';
import { QueryBuilder } from './query-builder';
import { AuthClient } from '../auth/index';
import { RealtimeChannel } from '../realtime/index';

/**
 * Configuration for HeliosDB client
 */
export interface HeliosDBConfig {
  /** Server URL */
  url: string;
  /** API key for authentication */
  apiKey?: string;
  /** JWT token for authentication */
  jwtToken?: string;
  /** Connection timeout in ms */
  connectTimeout?: number;
  /** Read timeout in ms */
  readTimeout?: number;
  /** Default branch */
  defaultBranch?: string;
  /** Extra headers */
  headers?: Record<string, string>;
}

/**
 * Internal query result implementation
 */
class QueryResultImpl<T = Record<string, unknown>> implements QueryResult<T> {
  columns: Column[];
  rows: unknown[][];
  rowCount: number;
  executionTimeMs: number;

  constructor(data: {
    columns: Column[];
    rows: unknown[][];
    row_count: number;
    execution_time_ms: number;
  }) {
    this.columns = data.columns;
    this.rows = data.rows;
    this.rowCount = data.row_count;
    this.executionTimeMs = data.execution_time_ms;
  }

  toObjects(): T[] {
    const colNames = this.columns.map((c) => c.name);
    return this.rows.map((row) => {
      const obj: Record<string, unknown> = {};
      for (let i = 0; i < colNames.length; i++) {
        obj[colNames[i]!] = row[i];
      }
      return obj as T;
    });
  }
}

/**
 * HeliosDB client for interacting with HeliosDB
 *
 * @example
 * ```typescript
 * const db = new HeliosDB({
 *   url: 'http://localhost:8080',
 *   apiKey: 'your-api-key'
 * });
 *
 * // Execute queries
 * const result = await db.query('SELECT * FROM users');
 * console.log(result.toObjects());
 *
 * // Vector operations
 * const store = db.vectorStore('embeddings', { dimension: 1536 });
 * await store.upsert([{ id: 'doc1', vector: [...], metadata: {} }]);
 * ```
 */
export class HeliosDB {
  private readonly config: Required<
    Pick<HeliosDBConfig, 'url' | 'connectTimeout' | 'readTimeout' | 'defaultBranch'>
  > &
    HeliosDBConfig;

  constructor(config: HeliosDBConfig) {
    this.config = {
      ...config,
      url: config.url.replace(/\/$/, ''),
      connectTimeout: config.connectTimeout ?? 10000,
      readTimeout: config.readTimeout ?? 30000,
      defaultBranch: config.defaultBranch ?? 'main',
    };
  }

  /**
   * Create client from environment variables
   */
  static fromEnv(): HeliosDB {
    const url = process.env.HELIOSDB_URL ?? 'http://localhost:8080';
    const apiKey = process.env.HELIOSDB_API_KEY;
    const jwtToken = process.env.HELIOSDB_JWT_TOKEN;
    const defaultBranch = process.env.HELIOSDB_DEFAULT_BRANCH ?? 'main';

    return new HeliosDB({ url, apiKey, jwtToken, defaultBranch });
  }

  /**
   * Build request headers
   */
  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'User-Agent': 'heliosdb-typescript/2.6.0',
    };

    if (this.config.apiKey) {
      headers['X-API-Key'] = this.config.apiKey;
    }
    if (this.config.jwtToken) {
      headers['Authorization'] = `Bearer ${this.config.jwtToken}`;
    }

    return { ...headers, ...this.config.headers };
  }

  /**
   * Make HTTP request
   */
  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    params?: Record<string, string>
  ): Promise<T> {
    let url = `${this.config.url}${path}`;

    if (params) {
      const searchParams = new URLSearchParams(params);
      url += `?${searchParams.toString()}`;
    }

    const response = await request(url, {
      method,
      headers: this.buildHeaders(),
      body: body ? JSON.stringify(body) : undefined,
      bodyTimeout: this.config.readTimeout,
      headersTimeout: this.config.connectTimeout,
    });

    const data = (await response.body.json()) as Record<string, unknown>;

    if (response.statusCode >= 400) {
      throw createErrorFromResponse(response.statusCode, data);
    }

    return data as T;
  }

  // ==========================================================================
  // Health & Info
  // ==========================================================================

  /**
   * Check server health
   */
  async health(): Promise<HealthStatus> {
    const data = await this.request<{
      status: 'healthy' | 'degraded' | 'unhealthy';
      version: string;
      uptime_seconds: number;
    }>('GET', '/health');

    return {
      status: data.status,
      version: data.version,
      uptimeSeconds: data.uptime_seconds,
    };
  }

  // ==========================================================================
  // Query Execution
  // ==========================================================================

  /**
   * Execute a SQL query
   *
   * @param sql - SQL query string
   * @param params - Query parameters
   * @param options - Query options
   * @returns Query results
   *
   * @example
   * ```typescript
   * const result = await db.query<User>(
   *   'SELECT * FROM users WHERE active = $1',
   *   [true]
   * );
   * for (const user of result.toObjects()) {
   *   console.log(user.name);
   * }
   * ```
   */
  async query<T = Record<string, unknown>>(
    sql: string,
    params?: unknown[],
    options?: QueryOptions
  ): Promise<QueryResult<T>> {
    const branch = options?.branch ?? this.config.defaultBranch;

    const data = await this.request<{
      columns: Column[];
      rows: unknown[][];
      row_count: number;
      execution_time_ms: number;
    }>('POST', `/v1/branches/${branch}/query`, {
      sql,
      params: params ?? [],
      timeout_ms: options?.timeoutMs ?? 30000,
      mode: options?.mode ?? 'normal',
    });

    return new QueryResultImpl<T>(data);
  }

  /**
   * Execute a SQL statement (INSERT, UPDATE, DELETE)
   *
   * @param sql - SQL statement
   * @param params - Query parameters
   * @param options - Query options
   * @returns Number of affected rows
   */
  async execute(sql: string, params?: unknown[], options?: QueryOptions): Promise<number> {
    const result = await this.query(sql, params, options);
    return result.rowCount;
  }

  // ==========================================================================
  // Branch Management
  // ==========================================================================

  /**
   * List all branches
   */
  async listBranches(): Promise<Branch[]> {
    const data = await this.request<{
      branches: Array<{
        name: string;
        parent: string | null;
        created_at: string;
        commit_count: number;
      }>;
    }>('GET', '/v1/branches');

    return data.branches.map((b) => ({
      name: b.name,
      parent: b.parent,
      createdAt: new Date(b.created_at),
      commitCount: b.commit_count,
    }));
  }

  /**
   * Get branch details
   */
  async getBranch(name: string): Promise<Branch> {
    const data = await this.request<{
      name: string;
      parent: string | null;
      created_at: string;
      commit_count: number;
    }>('GET', `/v1/branches/${name}`);

    return {
      name: data.name,
      parent: data.parent,
      createdAt: new Date(data.created_at),
      commitCount: data.commit_count,
    };
  }

  /**
   * Create a new branch
   */
  async createBranch(name: string, options?: CreateBranchOptions): Promise<Branch> {
    const data = await this.request<{
      name: string;
      parent: string | null;
      created_at: string;
      commit_count: number;
    }>('POST', '/v1/branches', {
      name,
      from_branch: options?.fromBranch ?? 'main',
      at_timestamp: options?.atTimestamp,
    });

    return {
      name: data.name,
      parent: data.parent,
      createdAt: new Date(data.created_at),
      commitCount: data.commit_count,
    };
  }

  /**
   * Delete a branch
   */
  async deleteBranch(name: string): Promise<void> {
    if (name === 'main') {
      throw new ValidationError('Cannot delete main branch');
    }
    await this.request('DELETE', `/v1/branches/${name}`);
  }

  /**
   * Use a branch with automatic cleanup
   *
   * @example
   * ```typescript
   * await db.withBranch('experiment-1', async (branch) => {
   *   await branch.execute("UPDATE config SET value = 'test'");
   *   // Branch is auto-deleted on exit
   * });
   * ```
   */
  async withBranch<T>(
    name: string,
    fn: (branch: BranchContext) => Promise<T>,
    options?: { fromBranch?: string; autoCleanup?: boolean }
  ): Promise<T> {
    const fromBranch = options?.fromBranch ?? 'main';
    const autoCleanup = options?.autoCleanup ?? true;

    await this.createBranch(name, { fromBranch });
    const ctx = new BranchContext(this, name, fromBranch);

    try {
      return await fn(ctx);
    } finally {
      if (autoCleanup && !ctx.merged) {
        try {
          await this.deleteBranch(name);
        } catch {
          // Best effort cleanup
        }
      }
    }
  }

  /**
   * Get a branch accessor for fluent API
   *
   * @example
   * ```typescript
   * const users = await db.branches('main')
   *   .table('users')
   *   .select(['id', 'name'])
   *   .where({ active: true })
   *   .limit(10)
   *   .execute();
   * ```
   */
  branches(name: string): BranchAccessor {
    return new BranchAccessor(this, name);
  }

  // ==========================================================================
  // Table Operations
  // ==========================================================================

  /**
   * List tables in a branch
   */
  async listTables(branch?: string): Promise<TableInfo[]> {
    const b = branch ?? this.config.defaultBranch;
    const data = await this.request<{
      tables: Array<{
        name: string;
        row_count: number;
        size_bytes: number;
        created_at: string;
      }>;
    }>('GET', `/v1/branches/${b}/tables`);

    return data.tables.map((t) => ({
      name: t.name,
      rowCount: t.row_count,
      sizeBytes: t.size_bytes,
      createdAt: new Date(t.created_at),
    }));
  }

  /**
   * Get table schema
   */
  async getTableSchema(table: string, branch?: string): Promise<TableSchema> {
    const b = branch ?? this.config.defaultBranch;
    const data = await this.request<{
      name: string;
      columns: Array<{
        name: string;
        type: string;
        nullable?: boolean;
        default?: string;
      }>;
      primary_key: string[];
      indexes: Array<{
        name: string;
        columns: string[];
        type: string;
        unique: boolean;
      }>;
    }>('GET', `/v1/branches/${b}/tables/${table}`);

    return {
      name: data.name,
      columns: data.columns,
      primaryKey: data.primary_key,
      indexes: data.indexes.map((idx) => ({
        name: idx.name,
        columns: idx.columns,
        type: idx.type as 'btree' | 'hash' | 'hnsw' | 'ivfflat',
        unique: idx.unique,
      })),
    };
  }

  // ==========================================================================
  // Data Operations
  // ==========================================================================

  /**
   * Insert rows into a table
   */
  async insert(
    table: string,
    rows: Record<string, unknown>[],
    options?: InsertOptions
  ): Promise<number> {
    const branch = options?.branch ?? this.config.defaultBranch;

    const data = await this.request<{ inserted: number }>(
      'POST',
      `/v1/branches/${branch}/tables/${table}/data`,
      {
        rows,
        on_conflict: options?.onConflict ?? 'error',
      },
      { auto_create: String(options?.autoCreate ?? false) }
    );

    return data.inserted;
  }

  /**
   * Infer schema from data
   */
  async inferSchema(
    data: Record<string, unknown>[],
    tableName?: string
  ): Promise<InferredSchema> {
    const result = await this.request<{
      inferred_schema: Record<string, string>;
      create_sql: string;
      confidence: number;
    }>('POST', '/v1/schema/infer', {
      data,
      table_name: tableName ?? 'inferred',
    });

    return {
      inferredSchema: result.inferred_schema,
      createSql: result.create_sql,
      confidence: result.confidence,
    };
  }

  // ==========================================================================
  // Vector Store
  // ==========================================================================

  /**
   * Get or create a vector store
   *
   * @example
   * ```typescript
   * const store = db.vectorStore('documents', { dimension: 1536 });
   *
   * await store.upsert([
   *   { id: 'doc1', vector: [...], metadata: { title: 'Hello' } }
   * ]);
   *
   * const results = await store.search(queryVector, { topK: 10 });
   * ```
   */
  vectorStore(name: string, config?: VectorStoreConfig): VectorStore {
    return new VectorStore(this, name, config);
  }

  /**
   * Retrieve relevant documents for a RAG query.
   * Combines vector search with optional SQL filtering.
   *
   * @param query - The query text (will be embedded server-side)
   * @param collection - Name of the vector store collection
   * @param opts - Optional parameters
   * @param opts.k - Number of results to return (default 5)
   * @param opts.filter - Metadata filter for narrowing results
   * @returns Array of matching documents with content, score, and metadata
   *
   * @example
   * ```typescript
   * const results = await db.rag("What is HeliosDB?", "docs", { k: 3 });
   * for (const r of results) {
   *   console.log(`[${r.score.toFixed(2)}] ${r.content}`);
   * }
   * ```
   */
  async rag(
    query: string,
    collection: string,
    opts?: { k?: number; filter?: Record<string, any> }
  ): Promise<{ content: string; score: number; metadata: Record<string, any> }[]> {
    const store = this.vectorStore(collection);
    const results = await store.search(
      // Pass the query string as a vector placeholder -- the server
      // performs embedding when it receives a string via the REST API.
      query as any,
      {
        topK: opts?.k ?? 5,
        filter: opts?.filter,
        includeMetadata: true,
      }
    );

    return results.map((r) => ({
      content: (r.metadata?.text as string) ?? '',
      score: r.score,
      metadata: r.metadata ?? {},
    }));
  }

  // ==========================================================================
  // Agent Memory
  // ==========================================================================

  /**
   * Get an agent memory interface
   *
   * @example
   * ```typescript
   * const memory = db.agentMemory('session-123');
   *
   * await memory.addMessage('user', 'Hello!');
   * await memory.addMessage('assistant', 'Hi there!');
   *
   * const results = await memory.search('greeting', { topK: 5 });
   * ```
   */
  agentMemory(sessionId: string): AgentMemory {
    return new AgentMemory(this, sessionId);
  }

  // ==========================================================================
  // Time Travel
  // ==========================================================================

  /**
   * Execute a query at a historical timestamp
   */
  async timeTravelQuery<T = Record<string, unknown>>(
    sql: string,
    timestamp: string,
    params?: unknown[],
    options?: TimeTravelOptions
  ): Promise<QueryResult<T>> {
    const branch = options?.branch ?? this.config.defaultBranch;

    const data = await this.request<{
      columns: Column[];
      rows: unknown[][];
      row_count: number;
      execution_time_ms: number;
    }>('POST', `/v1/branches/${branch}/time-travel`, {
      sql,
      timestamp,
      params: params ?? [],
    });

    return new QueryResultImpl<T>(data);
  }

  // ==========================================================================
  // Supabase-Compatible API
  // ==========================================================================

  /**
   * Start a Supabase-style fluent query on a table
   *
   * @param table - Table name to query
   * @returns A chainable QueryBuilder instance
   *
   * @example
   * ```typescript
   * // Fluent query (Supabase-compatible)
   * const { data, error } = await db.from('users')
   *   .select('id, name, email')
   *   .eq('active', true)
   *   .order('created_at', { ascending: false })
   *   .limit(10);
   *
   * // Single row
   * const { data: user } = await db.from('users')
   *   .select('*')
   *   .eq('id', 42)
   *   .single();
   * ```
   */
  from(table: string): QueryBuilder {
    return new QueryBuilder(
      table,
      this.config.url,
      this.config.apiKey ?? ''
    );
  }

  /**
   * Get the authentication client
   *
   * @example
   * ```typescript
   * // Sign up
   * const result = await db.auth.signUp({ email: 'user@example.com', password: 'secret' });
   *
   * // Sign in
   * const session = await db.auth.signIn({ email: 'user@example.com', password: 'secret' });
   *
   * // Get current user
   * const user = await db.auth.getUser(session.access_token);
   *
   * // Sign out
   * await db.auth.signOut();
   * ```
   */
  get auth(): AuthClient {
    return new AuthClient(this.config.url, this.config.apiKey ?? '');
  }

  /**
   * Create a realtime channel for change notifications
   *
   * @param topic - Table name or channel topic to subscribe to
   * @returns A RealtimeChannel that can be configured and subscribed
   *
   * @example
   * ```typescript
   * const channel = db.channel('users');
   *
   * channel
   *   .on('INSERT', (payload) => console.log('New row:', payload))
   *   .on('UPDATE', (payload) => console.log('Updated:', payload))
   *   .on('*', (payload) => console.log('Any change:', payload))
   *   .subscribe();
   *
   * // Later, disconnect
   * channel.unsubscribe();
   * ```
   */
  channel(topic: string): RealtimeChannel {
    const wsUrl = this.config.url
      .replace(/^http:/, 'ws:')
      .replace(/^https:/, 'wss:');
    return new RealtimeChannel(`${wsUrl}/realtime/v1/websocket`, topic);
  }

  // Internal request method for use by other classes
  /** @internal */
  _request<T>(
    method: string,
    path: string,
    body?: unknown,
    params?: Record<string, string>
  ): Promise<T> {
    return this.request(method, path, body, params);
  }

  /** @internal */
  get _defaultBranch(): string {
    return this.config.defaultBranch;
  }
}

/**
 * Branch accessor for fluent API
 */
class BranchAccessor {
  constructor(
    private readonly client: HeliosDB,
    readonly name: string
  ) {}

  /**
   * Start a fluent query on a table
   */
  table(name: string): TableQuery {
    return new TableQuery(this.client, this.name, name);
  }

  /**
   * Execute a query on this branch
   */
  query<T = Record<string, unknown>>(
    sql: string,
    params?: unknown[]
  ): Promise<QueryResult<T>> {
    return this.client.query<T>(sql, params, { branch: this.name });
  }

  /**
   * Execute a statement on this branch
   */
  execute(sql: string, params?: unknown[]): Promise<number> {
    return this.client.execute(sql, params, { branch: this.name });
  }
}
